src/amo/sagas/collections.js

```diff
@@ -166,43 +171,71 @@ export function* addAddonToCollection({
   }
 }
 
-export function* updateCollection({
+export function* modifyCollection({
+  type,
   payload: {
-    errorHandlerId,
     collectionSlug,
     defaultLocale,
     description,
+    errorHandlerId,
     formOverlayId,
     name,
     slug,
     user,
   },
-}: UpdateCollectionAction): Generator<any, any, any> {
+}: {|
+  type: typeof CREATE_COLLECTION | typeof UPDATE_COLLECTION,
+  payload: {
+    ...RequiredModifyCollectionParams,
+    ...OptionalModifyCollectionParams,
+    collectionSlug?: string,
+    name?: ?LocalizedString,
+    slug?: ?string,
+  }
+|}): Generator<any, any, any> {
+  const creating = type === CREATE_COLLECTION;
   const errorHandler = createErrorHandler(errorHandlerId);
   yield put(errorHandler.createClearingAction());
 
   try {
     yield put(beginFormOverlaySubmit(formOverlayId));
 
     const state = yield select(getState);
-    const params: UpdateCollectionParams = {
-      api: state.api,
-      collectionSlug,
-      defaultLocale,
-      description,
-      name,
-      slug,
-      user,
-    };
-    yield call(api.updateCollection, params);
 
-    const slugWasEdited = slug && slug !== collectionSlug;
-    const effectiveSlug = slug || collectionSlug;
+    if (creating) {
+      invariant(name, 'name cannot be empty when creating');
+      invariant(slug, 'slug cannot be empty when creating');
+      const apiParams: CreateCollectionParams = {
+        api: state.api,
+        defaultLocale,
+        description,
+        name,
+        slug,
+        user,
+      };
+      yield call(api.createCollection, apiParams);
+    } else {
+      invariant(collectionSlug,
+        'collectionSlug cannot be empty when updating');
+      const apiParams: UpdateCollectionParams = {
```

Comment:
Hmm. I see a lot of duplication between these two param objects. You can introduce a spread with a shared base object but Flow isn't able to compute all spread properties so it will complain about exact object types. There is a global `$Shape` utility we use to turn an exact type into an inexact type. It's not great because you lose the type safety around unexpected properties but being able to DRY up this code seems better.

Something like this should work as long you can get the disjoint unions to kick in:

```diff
diff --git a/src/amo/sagas/collections.js b/src/amo/sagas/collections.js
index 5cf55a0b6..f48a5a31f 100644
--- a/src/amo/sagas/collections.js
+++ b/src/amo/sagas/collections.js
@@ -202,29 +202,26 @@ export function* modifyCollection({
 
     const state = yield select(getState);
 
+    const baseApiParams = {
+      api: state.api,
+      defaultLocale,
+      description,
+      name,
+      slug,
+      user,
+    };
+
     if (creating) {
       invariant(name, 'name cannot be empty when creating');
       invariant(slug, 'slug cannot be empty when creating');
-      const apiParams: CreateCollectionParams = {
-        api: state.api,
-        defaultLocale,
-        description,
-        name,
-        slug,
-        user,
-      };
+      const apiParams: $Shape<CreateCollectionParams> = { ...baseApiParams };
       yield call(api.createCollection, apiParams);
     } else {
       invariant(collectionSlug,
         'collectionSlug cannot be empty when updating');
-      const apiParams: UpdateCollectionParams = {
-        api: state.api,
+      const apiParams: $Shape<UpdateCollectionParams> = {
+        ...baseApiParams,
         collectionSlug,
-        defaultLocale,
-        description,
-        name,
-        slug,
-        user,
       };
       yield call(api.updateCollection, apiParams);
     }
```

If this turns into a headache then it's no big deal. The most important part is gaining type safety for the API calls, which you already had.